File: 5.longest-palindromic-substring.cpp
/*
 * @lc app=leetcode id=5 lang=cpp
 *
 * [5] Longest Palindromic Substring
 */
#include <string>
#include <vector>
using namespace std;
class Solution
{
public:
	string longestPalindrome(string s)
	{
		if (s.empty()) return "";
		int length = s.length();
		if (length == 1) return s;
		bool** dp = new bool*[length];
		for (int i = 0; i < length; ++i)
		{
			dp[i] = new bool[length - i]{false};
		}
		int l = 1;
		int index = 0;
		for (int len = 0; len < length; ++len)
		{
			for (int i = 0; i + len < length; ++i)
			{
				if (len == 0)
				{
					dp[i][len] = true;
				}
				else if (len == 1)
				{
					if (s[i] == s[i + len])
					{
						l = 2;
						index = i;
						dp[i][len] = true;
					}
				}
				else
				{
					if (s[i] == s[i + len] && dp[i + 1][len - 2])
					{
						l = len + 1;
						index = i;
						dp[i][len] = true;
					}
				}
			}
		}
		return s.substr(index, l);
	}
};

File: 5.longest-palindromic-substring.java
/*
 * @lc app=leetcode id=5 lang=java
 *
 * [5] Longest Palindromic Substring
 *
 * https://leetcode.com/problems/longest-palindromic-substring/description/
 *
 * algorithms
 * Medium (27.96%)
 * Likes:    5889
 * Dislikes: 486
 * Total Accepted:    846.8K
 * Total Submissions: 2.9M
 * Testcase Example:  '"babad"'
 *
 * Given a string s, find the longest palindromic substring in s. You may
 * assume that the maximum length of s is 1000.
 * 
 * Example 1:
 * 
 * 
 * Input: "babad"
 * Output: "bab"
 * Note: "aba" is also a valid answer.
 * 
 * 
 * Example 2:
 * 
 * 
 * Input: "cbbd"
 * Output: "bb"
 * 
 * 
 */

// @lc code=start
class Solution {
    public String longestPalindrome(String s) {
        if (s == null || s.length() == 0)
            return "";
        if (s.length() == 1)
            return s;
        boolean[][] dp = new boolean[s.length()][];
        for (int i = 0; i < dp.length; i++) {
            dp[i] = new boolean[s.length() - i + 1];
            dp[i][0] = true;
            dp[i][1] = true;
        }
        int maxlength = 1;
        int maxBegin = 0;
        for (int len = 2; len <= s.length(); len++)
            for (int i = 0; i <= s.length() - len; i++) {
                if (s.charAt(i) == s.charAt(i + len - 1) && dp[i + 1][len - 2]) {
                    maxlength = len;
                    maxBegin = i;
                    dp[i][len] = true;
                } else
                    dp[i][len] = false;
            }
        return s.substring(maxBegin, maxBegin + maxlength);
    }
}
// @lc code=end
